File: 0080-remove-duplicates-from-sorted-array-ii/0080-remove-duplicates-from-sorted-array-ii.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        before = [0, 0]
        ref = 0
        k = 0
        for i in range(len(nums)):
            if before[0] != nums[i] or i == 0:
                nums[ref] = nums[i]
                before = [nums[i], 1]
                k += 1
                ref += 1
            elif before[0] == nums[i] and before[1] < 2:
                nums[ref] = nums[i]
                before[1] += 1
                k += 1
                ref += 1
        return k

File: 0080-remove-duplicates-from-sorted-array-ii/0080-remove-duplicates-from-sorted-array-ii.java
class Solution {
    public int removeDuplicates(int[] nums) {
        int n = nums.length;

        if (n < 3)
            return n;

        int k = 1;
        int i = 1;
        int before = nums[0];
        int cnt = 1;

        while (i < n) {
            if (nums[i] == before) {
                cnt++;
                if (cnt == 3) {
                    while (i < n && nums[i] == before) {
                        i++;
                    }

                    if (i == n)
                        break;

                    before = nums[i];
                    nums[k] = nums[i];
                    cnt = 1;
                    k++;
                    i++;
                } else {
                    nums[k] = nums[i];
                    k++;
                    i++;
                }
            } else {
                nums[k] = nums[i];
                before = nums[i];
                cnt = 1;
                i++;
                k++;
            }
        }

        return k;
    }
}